offchain-modules/packages/x/src/handlers/eth.ts

```diff
@@ -1,63 +1,191 @@
-        );
-        logger.debug('EthHandler watchLockEvents eth lockEvtLog:', { log, parsedLog });
-        const amount = parsedLog.args.lockedAmount.toString();
-        if (amount === '0') {
-          return;
-        }
-        if (this.role === 'collector') {
-          await this.db.createCkbMint([
-            {
-              id: log.transactionHash,
-              chain: ChainType.ETH,
-              amount: amount,
-              asset: parsedLog.args.token,
-              recipientLockscript: uint8ArrayToString(fromHexString(parsedLog.args.recipientLockscript)),
-              sudtExtraData: parsedLog.args.sudtExtraData,
-            },
-          ]);
-          logger.info(`EthHandler watchLockEvents save CkbMint successful for eth tx ${log.transactionHash}.`);
+  async getLastHandledBlock(): Promise<{ blockNumber: number; blockHash: string }> {
+    const lastHandledBlock = await this.kvDb.get(lastHandleEthBlockKey);
+    if (!lastHandledBlock || lastHandledBlock.length === 0) {
+      return { blockNumber: 0, blockHash: '' };
+    }
+    const block = lastHandledBlock[0].value.split(',');
+    return { blockNumber: parseInt(block[0]), blockHash: block[1] };
+  }
+
+  async setLastHandledBlock(blockNumber: number, blockHash: string): Promise<void> {
+    this.lastHandledBlockHeight = blockNumber;
+    this.lastHandledBlockHash = blockHash;
+    await this.kvDb.set(lastHandleEthBlockKey, `${blockNumber},${blockHash}`);
+  }
+
+  async init() {
+    const currentBlockHeight = await this.ethChain.getCurrentBlockNumber();
+    const lastHandledBlock = await this.getLastHandledBlock();
+    if (lastHandledBlock.blockNumber === 0) {
+      const currentBlock = await this.ethChain.getBlock(currentBlockHeight);
+      this.lastHandledBlockHash = currentBlock.hash;
+      this.lastHandledBlockHeight = currentBlock.number;
+    } else {
+      this.lastHandledBlockHash = lastHandledBlock.blockHash;
+      this.lastHandledBlockHeight = lastHandledBlock.blockNumber;
+    }
+
+    logger.info(
+      `EthHandler init lastHandledBlock:${this.lastHandledBlockHeight} currentBlockHeight: ${currentBlockHeight}`,
+    );
+
+    if (this.lastHandledBlockHeight === currentBlockHeight) {
+      return;
+    }
+    const confirmNumber = ForceBridgeCore.config.eth.confirmNumber;
+    const getLogBatchSize = 100;
+    for (;;) {
+      const currentBlockHeight = await this.ethChain.getCurrentBlockNumber();
+      if (currentBlockHeight === this.lastHandledBlockHeight) {
+        break;
+      }
+
+      for (;;) {
+        const endBlockNumber =
+          this.lastHandledBlockHeight + getLogBatchSize > currentBlockHeight
+            ? currentBlockHeight
+            : this.lastHandledBlockHeight + getLogBatchSize;
+
+        logger.info(`EthHandler init getLogs from:${this.lastHandledBlockHeight} to:${endBlockNumber}`);
+
+        await this.ethChain.getLogs(this.lastHandledBlockHeight, endBlockNumber - 1, async (log, parsedLog) => {
```

Comment:
It's weird to use handleFunc as param in `getLogs` function.
We can remove the param, just return logs in `getLogs` and handle it without callback style.

Reply:
Why `endBlockNumber - 1` here? It seems will miss the endBlock in this way.

Reply:
the bound of logs return by getLogs is [ and ], if getLogs(this.lastHandledBlockHeight, endBlockNumber), may get logs duplicate with the endBlock

Reply:
so I recommend `[this.lastHandledBlockHeight + 1, endBlockNumber]`, because you set lastHandledBlockHeight to endBlockNumber but actually not handle endBlockNumber.